stl/inc/xstring

```diff
@@ -4586,72 +4840,72 @@ private:
+        _ASAN_STRING_CREATE(*this);
         return *this;
     }
 
-    void _Become_small() {
+    _CONSTEXPR20 void _Become_small() {
         // release any held storage and return to small string mode
-        // pre: *this is in large string mode
-        // pre: this is small enough to return to small string mode
-        // (not _CONSTEXPR20; SSO should be disabled in a constexpr context)
+        auto& _My_data = _Mypair._Myval2;
+        _STL_INTERNAL_CHECK(_My_data._Large_string_engaged());
+        _STL_INTERNAL_CHECK(_My_data._Mysize < _BUF_SIZE);
 
-        _Mypair._Myval2._Orphan_all();
-        const pointer _Ptr = _Mypair._Myval2._Bx._Ptr;
+        _My_data._Orphan_all();
+        _ASAN_STRING_CLEAR(*this);
+        const pointer _Ptr = _My_data._Bx._Ptr;
         auto& _Al          = _Getal();
-        _Destroy_in_place(_Mypair._Myval2._Bx._Ptr);
-        _Traits::copy(_Mypair._Myval2._Bx._Buf, _Unfancy(_Ptr), _Mypair._Myval2._Mysize + 1);
-        _Al.deallocate(_Ptr, _Mypair._Myval2._Myres + 1);
-        _Mypair._Myval2._Myres = _BUF_SIZE - 1;
+        _Destroy_in_place(_My_data._Bx._Ptr);
+#if _HAS_CXX20
+        if (_STD is_constant_evaluated()) {
+            _Construct_in_place(_Mypair._Myval2._Bx);
+        }
+#endif // _HAS_CXX20
+        _Traits::copy(_My_data._Bx._Buf, _Unfancy(_Ptr), _My_data._Mysize + 1);
+        _Al.deallocate(_Ptr, _My_data._Myres + 1);
+        _My_data._Myres = _BUF_SIZE - 1;
+        _ASAN_STRING_CREATE(*this);
     }
 
     _CONSTEXPR20 void _Eos(const size_type _Newsize) { // set new length and null terminator
+        _ASAN_STRING_SHRINK(*this, _Mypair._Myval2._Mysize - _Newsize);
         _Traits::assign(_Mypair._Myval2._Myptr()[_Mypair._Myval2._Mysize = _Newsize], _Elem());
     }
 
     _CONSTEXPR20 void _Tidy_init() noexcept { // initialize basic_string data members
-        auto& _My_data   = _Mypair._Myval2;
-        _My_data._Mysize = 0;
+        auto& _My_data = _Mypair._Myval2;
 
+        _My_data._Mysize = 0;
+        _My_data._Myres  = _BUF_SIZE - 1;
+        // the _Traits::assign is last so the codegen doesn't think the char write can alias this
 #if _HAS_CXX20
         if (_STD is_constant_evaluated()) {
-            _My_data._Myres        = _BUF_SIZE; // SSO disabled in constexpr context
-            auto& _Al              = _Getal();
-            const pointer _New_ptr = _Al.allocate(_BUF_SIZE + 1); // throws
-            _My_data._Bx._Ptr      = _New_ptr;
-
-            _Elem* const _Raw_new = _Unfancy(_New_ptr);
-            _Traits::assign(_Raw_new, _BUF_SIZE + 1, _Elem());
-        } else
-#endif // _HAS_CXX20
-        {
-            _My_data._Myres = _BUF_SIZE - 1;
-            // the _Traits::assign is last so the codegen doesn't think the char write can alias this
-            _Traits::assign(_My_data._Bx._Buf[0], _Elem());
+            _Construct_in_place(_Mypair._Myval2._Bx);
         }
+#endif // _HAS_CXX20
+        _Traits::assign(_My_data._Bx._Buf[0], _Elem());

```

Comment:
Stupid ?: Do we need the _Construct_in_place any longer after Richard's Ninja-Patch? http://eel.is/c++draft/class.union#general-6

Reply:
That is an excellent question and was recently discussed in the reflector. 

TL;DR Yes we need them

The issue is that the ninja wording only applies to *assignments*. However, we must not perform and assignment but need to call `char_traits::assign`

The issue is that in order to call `char_traits::assign` we need the address / reference to a given element of the buffer. But without starting the lifetime of the element this is UB. 

So it could work for an SSO vector, but not for `string` with its indirection